### ISSUE TITLE
Dashed Lines and Outline Styling

### ISSUE THREAD

**Issue by JamesLMilner** [JamesLMilner/terra-draw#48]
This was originally raised in #38 but raising as a separate issue to more easily track it.

The purposed of this issue is to explore how viable it would be to support dash lines and outlines across mapping libraries. There is a somewhat philosophical question of if we we provide a fall back to a normal line if it is not supported and if we should have 'best effort' features in Terra Draw to allow support for more complex options.

**Comment by JamesLMilner** [comment]
I started looking into this - I got this working with Leaflet, OpenLayers and Mapbox, but for MapLibre it is only possible to hardcode the dash array styling as a literal value (i.e. it cannot be driven by data/properties). 

See https://github.com/maplibre/maplibre-gl-js/issues/1235 for more details. 

In theory we could push this forward without MapLibre support, but it feels weird to have one piece of styling functionality which does not work for one adapter. The other option is we do support MapLibre for dash lines but the value would be hardcoded to say `[1, 1]` and changing the styling property would not change it. 

If people have thoughts on this I would be open to hearing them about what might be there preference?

**Comment by JamesLMilner** [comment]
It looks like https://github.com/maplibre/maplibre-gl-js/issues/1235 has been picked up under it's bounty programme. Hoping we might be able to achieve this at some point next year!

**Comment by JamesLMilner** [comment]
Still no movement on this since it was picked up on the bounty programme. I will continue to keep an eye open!

**Comment by JamesLMilner** [comment]
This looks like it is being actively worked on 🎉 Exciting times: https://github.com/maplibre/maplibre-gl-js/pull/5812

**Comment by JamesLMilner** [comment]
I'm going to pin this a few different issues have raised in relation to this and want to make sure it's visible to people 👍🏻

**Comment by JamesLMilner** [comment]
This looks like it is now merged! https://github.com/maplibre/maplibre-gl-js/pull/5812 . It looks like 5.8.0 now supports t his from the changelog: https://github.com/maplibre/maplibre-gl-js/blob/main/CHANGELOG.md#580

**Comment by RLaipins** [comment]
👋🏻@JamesLMilner,  I'm currently on Maplibre creating a new drawing mode. I would like to have a dashed outline while I'm drawing. Has there been any progress on this issue or suggestions on how could It be possible have dashed outline?

**Comment by JamesLMilner** [comment]
@RLaipins there's an experimental branch for supporting dash lines in for LineStrings in all adapters here: https://github.com/JamesLMilner/terra-draw/tree/dash-line

When I get the time I will try and get it across the line, it just requires releasing each adapter so requires a bit of thought and coordination. 

Once that's in we can look at supporting them for polygon outlines too.

**Comment by JamesLMilner** [comment]
Just a note this has been delayed slightly due to #442 being resolved first (full opacity support for all rendered items) - it was the easier case and we wanted to get that covered before taking this further.

**Comment by TomBeckett** [comment]
@JamesLMilner also on Maplibre creating a new drawing mode. Definitely something we'd love to see!

**Comment by JamesLMilner** [comment]
I've completed work towards supporting dashed line strings. It's a bit more work to get this function with polygon outlines.

**Comment by JamesLMilner** [comment]
Polygon outlines are currently blocked by the fact that doing dashed outlines consistently around polygons in Google Maps is very tricky (potentially impossible in a sensible way?). [One solution](https://stackoverflow.com/questions/41967862/dashed-polygons-google-maps) appears to be to use a Polyline for the outline, but even this appears to be complicated to get consistent behaviour where the outline doesn't under or overshoot the perimeter of the polygon.

**Comment by RLaipins** [comment]
@JamesLMilner is google maps adapter the only one with this issue or there are other adapters blocking it?

> I started looking into this - I got this working with Leaflet, OpenLayers and Mapbox, but for MapLibre it is only possible to hardcode the dash array styling as a literal value (i.e. it cannot be driven by data/properties).
> 
> See [maplibre/maplibre-gl-js#1235](https://github.com/maplibre/maplibre-gl-js/issues/1235) for more details.
> 
> In theory we could push this forward without MapLibre support, but it feels weird to have one piece of styling functionality which does not work for one adapter. The other option is we do support MapLibre for dash lines but the value would be hardcoded to say `[1, 1]` and changing the styling property would not change it.
> 
> If people have thoughts on this I would be open to hearing them about what might be there preference?

^ I see that previously there was question about piece of styling being available for few adapters. I think this would be a good compromise (of course this would need some warnings inside docs).  

I noticed that line dash styling is missing from select mode and styling itself is static only. Is that on purpose?

**Comment by JamesLMilner** [comment]
@RLaipins generally I am hesitant to add features that can't be supported by all adapters. This is just a guiding principal that helps keep things simple. I understand that this must be frustrating for some users. In theory I think you could work around this with a custom adapter (i.e. vendoring in the MapLibre Adapter, then using custom properties to render the dashed outline for polygons).

Thank you for pointing out these two issues with line styling - I would infact classify these both as bugs and so have added a PR to address them! https://github.com/JamesLMilner/terra-draw/pull/923

**Comment by EduNauta** [comment]
Is a dotted line implementable?

Before this was implemented, i managed to vibe code a dash line tool, and also a dot line tool. 
Once you do one, doing the other should be trivial. It's just a different set of values for the longitude of the dash.
